js/style/style.js

```diff
@@ -455,13 +455,15 @@ class Style extends Evented {
 
         if (this._removedLayers[id] && layer.source) {
             // If, in the current batch, we have already removed this layer
-            // and we are now re-adding it, then we need to clear (rather
-            // than just reload) the underyling source's tiles.
-            // Otherwise, tiles marked 'reloading' will have buffers that are
-            // set up for the _previous_ version of this layer, confusing
+            // and we are now re-adding it with a different `type`, then we
+            // need to clear (rather than just reload) the underyling source's
+            // tiles.  Otherwise, tiles marked 'reloading' will have buckets /
+            // buffers that are set up for the _previous_ version of this
+            // layer, causing, e.g.:
             // https://github.com/mapbox/mapbox-gl-js/issues/3633
+            const removed = this._removedLayers[id];
             delete this._removedLayers[id];
-            this._updatedSources[layer.source] = 'clear';
+            this._updatedSources[layer.source] = removed.type !== layer.type ? 'clear' : 'reload';
```

Comment:
> although I think this actually points to a more general issue, which is that we do not handle adding and removing a layer within the same 'batch'

Ah, nevermind -- we handle this more general thing fine.

@lucaswoj the case you raise -- "remove a layer, add a new layer with a different type, remove that layer, add a new layer with the original type" -- would, indeed, cause a flash as in the original issue here.  However, I'd be inclined to punt on this, since it's a pretty extreme edge case that I do not expect would actually appear in the wild, and even when it does occur, the symptom is not particularly severe.